main/src/mill/main/MainModule.scala

```diff
@@ -177,4 +182,39 @@ trait MainModule extends mill.Module{
       }
     }
   }
+
+  /**
+    * Deletes the given targets from the out directory. Providing no targets
+    * will clean everything.
+    */
+  def clean(evaluator: Evaluator[Any], targets: String*) = mill.T.command {
+    val rootDir = ammonite.ops.pwd / OutDir
+
+    val KeepPattern = "(mill-.+)".r.anchored
```

Comment:
Here's an idea: what if instead of this regex, we simply treated `mill clean` as the same as `mill clean __`? That should correctly delete everything that needs to be deleted while leaving the not-to-be-deleted files untouched

Reply:
You jut beat me to it! I was starting to think along those line, mainly for better consistency with the other tasks in the main module, but you're right that it should also remove the special-case code with the somewhat arbitrary regex. 👌

I'll try to implement that and see if there is any issue.

Reply:
@guilgaly do you want to try out this one in this PR?

Reply:
@rockjam I'd say this PR is already useful, so I'll rather try to write the improved version for another future PR.

@lihaoyi & @rockjam : one thing I'm wondering about is wether we have a way to list all available external modules (like the GenIdea one)? As we'll want to clean their outputs too when doing `mill clean __`.

Reply:
I don't think we have a way to list all of them right now; they just live arbitrarily on the classpath, and JVM classpath-scanning is not well supported (slow, unreliable, ...). Currently, their target directories are also randomly scattered throughout the `out/` folder.

Perhaps we could just put all the `ExternalModule`s into an `mill-external` folder, and add that folder to the hardcoded list of things to `rm`?